INTYGFV-16493: Added prognos to FK7443

mhornfeldt opened the pull request.

mhornfeldt commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificatemodel/model/DateRangeType.java:
```diff
@@ -0,0 +1,39 @@
+  HELT_NEDSATT("100 procent");
+
+  private final String label;
+  public static final String CODE_SYSTEM = "KV_FKMU_0003";
```
This is taken from lisjp, maybe it's a different one

mhornfeldt commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificatemodel/model/DateRangeType.java:
```diff
@@ -0,0 +1,39 @@
+import lombok.Getter;
+
+@Getter
+public enum DateRangeType {
```
Better name for this?

ChrilleSigerhed replied:
I suggest WorkCapacityType

mhornfeldt commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/dto/config/CheckboxDateRangeConfig.java:
```diff
@@ -0,0 +1,39 @@
+@JsonDeserialize(builder = CheckboxDateRangeConfigBuilder.class)
+@Value
+@Builder
+public class CheckboxDateRangeConfig {
```
Better name for this?

ChrilleSigerhed replied:
I think this name works fine since its beeing used in CertificateDataConfigCheckboxDateRangeList.

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/dto/config/CertificateDataConfigCheckboxDateRangeList.java:
```diff
@@ -0,0 +1,56 @@
+/*
```
Remove licence code

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/dto/value/CertificateDataValueDateRange.java:
```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

mhornfeldt replied:
Since I can search for where it's used you can add this comment once instead of every occasion. It gets hard to see which comments actually touches code and not the license.

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/dto/value/CertificateDataValueDateRangeList.java:
```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/dto/value/CertificateDataValueType.java:
```diff
@@ -19,5 +19,5 @@
 
 public enum CertificateDataValueType {
-  DATE, TEXT
+  DATE, TEXT, DATE_RANGE, DATE_RANGE_LIST
```
Remove licence code here aswell please.

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/application/certificate/service/converter/CertificateDataCheckboxDateRangeListConfigConverter.java:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/infrastructure/certificatemodel/CertificateElementRuleFactory.java:
```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/infrastructure/certificatemodel/CertificateElementRuleFactory.java:
```diff
@@ -0,0 +1,74 @@
+    throw new IllegalStateException("Utility class");
+  }
+
+  public static ElementRule mandatory(ElementId id, FieldId fieldId) {
```
Nice! 
Lets utilize this factory method for FK7211 mandatory question aswell!

Should we add a factory method for ElementRuleLimit aswell?

mhornfeldt replied:
Done

ChrilleSigerhed commented on certificate-service/app/src/main/java/se/inera/intyg/certificateservice/infrastructure/certificatemodel/fk7443/CertificateModelFactoryFK7443.java:
```diff
@@ -129,24 +144,36 @@ private static ElementSpecification categorySymptom(
+                    QUESTION_SYMPTOM_ID,
+                    QUESTION_SYMPTOM_FIELD_ID
+                ),
                 ElementRuleLimit.builder()
```
See comment regarding factory method for limit

ChrilleSigerhed commented on certificate-service/app/src/test/java/se/inera/intyg/certificateservice/application/certificate/service/converter/CertificateDataCheckboxDateRangeListConfigConverterTest.java:
```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificate/model/ElementValueDateRangeList.java:
```diff
@@ -0,0 +1,36 @@
+@Builder
+public class ElementValueDateRangeList implements ElementValue {
+
+  FieldId dateRangeListId; // TODO: This cant be converted because wc doesnt have id for dr list
```
Remove comment

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificate/model/DateRange.java:
```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificate/model/DateRange.java:
```diff
@@ -0,0 +1,37 @@
+public class DateRange {
+
+  FieldId dateRangeId;
+  @With
```
Can be removed? Dont see any usage of this

mhornfeldt replied:
Done

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificate/model/ElementValueDateRangeList.java:
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificatemodel/model/CheckboxDateRange.java:
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificatemodel/model/DateRangeType.java:
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/certificatemodel/model/ElementConfigurationCheckboxDateRangeList.java:
```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/validation/model/ElementValidationDateRangeList.java:
```diff
@@ -0,0 +1,210 @@
+/*
+ * Copyright (C) 2024 Inera AB (http://www.inera.se)
```
Remove licence code

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/validation/model/ElementValidationDateRangeList.java:
```diff
@@ -0,0 +1,210 @@
+  }
+
+  private ElementValueDateRangeList getValue(ElementValue value) {
+    if (value == null) {
```
Redundant check since its already done earlier in the code

mhornfeldt replied:
Data is checked if it's null before but not value. This is the pattern we followed with the other validation. Which line is it that you mean it's checked before?

ChrilleSigerhed replied:
My bad, keep this as it is! :)

ChrilleSigerhed commented on certificate-service/domain/src/main/java/se/inera/intyg/certificateservice/domain/validation/model/ElementValidationDateRangeList.java:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    final var incompleteErrors = getIncompleteDateRangeErrors(data, categoryId, dateRangeList);
+    final var incorrectErrors = getIncorrectDateRangeErrors(data, categoryId, dateRangeList);
```
Maybe we should change the name of this method to better communicate that we are checking if the start date is after the end date?

mhornfeldt replied:
Done

ChrilleSigerhed reviewed (changes requested):
Very well done, espessially impressive how you handled the validation of the new component.

Overall the code is very easy to follow and clean, great usage of the newly estabilished pattern to define new converters for new types.

Left some minor comments to consider.

mhornfeldt commented:
@ChrilleSigerhed I removed the license headers, the comment and the with annotation. I expanded the rule factory and its usage. I renamed the incorrectDateRange errors variable.
